lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,421 +25,546 @@
+
+        if (!visited.contains(currentNode)) {
+          visited.insert(currentNode);
+          onStack[currentNode] = true;
+          LLVM_DEBUG(llvm::dbgs()
+                     << "\n visiting :"
+                     << reachingDefGraph[currentNode].first.getValue().getType()
+                     << reachingDefGraph[currentNode].first.getValue() << ","
+                     << reachingDefGraph[currentNode].first.getFieldID() << "\n"
+                     << getName(reachingDefGraph[currentNode].first));
+
+          FieldRef currentF = reachingDefGraph[currentNode].first;
+          if (recordPortPaths && currentNode != rootNode) {
+            if (isa<mlir::BlockArgument>(currentF.getValue()))
+              portPaths[reachingDefGraph[rootNode].first].insert(currentF);
+            // Even if the current node is not a port, there can be RWProbes of
+            // the current node at the port.
+            addToPortPathsIfRWProbe(
+                currentNode, portPaths[reachingDefGraph[rootNode].first]);
+          }
+        } else {
+          onStack[currentNode] = false;
+          dfsStack.pop_back();
+        }
+
+        for (auto neighbor : graph[currentNode].second) {
+          if (!visited.contains(neighbor)) {
+            dfsStack.push_back(neighbor);
+          } else if (onStack[neighbor]) {
+            // Cycle found !!
+            SmallVector<FieldRef, 16> path;
+            auto loopNode = neighbor;
+            // Construct the cyclic path.
+            do {
+              SmallVector<unsigned>::iterator it =
+                  llvm::find_if(reachingDefGraph[loopNode].second,
+                                [&](unsigned node) { return onStack[node]; });
+              if (it == reachingDefGraph[loopNode].second.end())
+                break;
+
+              path.push_back(reachingDefGraph[loopNode].first);
+              loopNode = *it;
+            } while (loopNode != neighbor);
+
+            reportLoopFound(path, reachingDefGraph[neighbor].first.getLoc());
+            return failure();
+          }
+        }
+      }
+      return success();
     };
-    auto errorDiag = mlir::emitError(
-        module.getLoc(), "detected combinational cycle in a FIRRTL module");
 
-    SmallVector<Value, 16> path;
-    path.push_back(childVal);
-    visiting.popUntilVal(
-        childVal, [&](Value visitingVal) { path.push_back(visitingVal); });
-    assert(path.back() == childVal);
-    path.pop_back();
+    DenseSet<unsigned> visited;
+    for (unsigned node = 0; node < graph.size(); ++node) {
+      bool isPort = false;
+      if (auto arg =
+              dyn_cast<BlockArgument>(reachingDefGraph[node].first.getValue()))
+        if (module.getPortDirection(arg.getArgNumber()) == Direction::Out) {
+          // For output ports, reset the visited. Required to revisit the entire
+          // graph, to discover all the paths that exist from any input port.
+          visited.clear();
```

Comment:
Not sure I follow, but maybe this is for handling module-local /reads/ of output ports, and possible cycles involving that? (or something else?)

Reply:
Oh, okay sure :+1: .  Thanks for the explanation!